src/stores/pacbioPools.js

```diff
@@ -0,0 +1,110 @@
+import { defineStore } from 'pinia'
+import useRootStore from '@/stores'
+import { handleResponse } from '@/api/ResponseHelper.js'
+import { groupIncludedByResource, dataToObjectById } from '@/api/JsonApi.js'
+
+/**
+ * This function takes an object with `pool` and `used_aliquots` properties and returns an array of run suitability errors.
+ * It maps over the errors of `pool.run_suitability` and `used_aliquot.run_suitability` for each used_aliquot, formats the errors with the pool or used_aliquot details, and returns the formatted errors.
+ *
+ * @param {Object}  - An object with `pool` and `used_aliquots` properties.
+ * @returns {string[]} The formatted run suitability errors.
+ */
+const buildRunSuitabilityErrors = ({ pool, used_aliquots }) => [
+  ...pool.run_suitability.errors.map(({ detail }) => `Pool ${detail}`),
+  ...used_aliquots.flatMap((used_aliquot) => {
+    const used_aliquotName = `Used aliquot ${used_aliquot.id} (${used_aliquot.sample_name})`
+    return used_aliquot.run_suitability.errors.map(({ detail }) => `${used_aliquotName} ${detail}`)
+  }),
+]
+/**
+ * This store manages the state of PacBio pools which are fetched from the API and used in the PacBio pools page table.
+ * It contains the pools, tubes, used_aliquots, requests, and tags state properties, and the fetchPools action.
+ * @exports usePacbioPools
+ */
+export const usePacbioPoolsStore = defineStore('pacbioPools', {
+  state: () => ({
+    pools: {},
+    tubes: {},
+    used_aliquots: {},
+    requests: {},
+    tags: {},
+  }),
+  getters: {
+    /**
+     * This function takes a `state` object and returns an array of pools.
+     * It maps over the values of `state.pools`, retrieves the used_aliquots, barcode, and run suitability for each pool, and
+     * returns the pools with the retrieved data.
+     * The used_aliquots for each pool are retrieved by mapping over the `pool.used_aliquots`, retrieving the used_aliquot data from `state.used_aliquots`
+     * based on the used_aliquot ID, and returning the used_aliquot data with the sample name from `state.requests` and the group ID from `state.tags`.
+     * The barcode for each pool is retrieved from `state.tubes` based on the `pool.tube`.
+     * The run suitability for each pool is retrieved from `pool.run_suitability` and formatted with the `buildRunSuitabilityErrors` function.
+     *
+     * @param {Object} state - The state object.
+     * @returns {Object[]} The array of pools with the retrieved data.
+     */
+    poolsArray: (state) => {
```

Comment:
It looks like we are doing objectArray in a lot of places. There might be an opportunity for abstraction somewhere.

Reply:
I understand but it is something worth thinking about. Javascript is a functional language so is built for stuff like this. Sounds like we are formatting all fetched data to a shape that components can access so we can pass functions to do certain parts of it.